juspay/hyperswitch
# feat(router): add payments manual-update api

## Type of Change



## Description
add payments manual-update api

1. manual-update api allows to update the attempt status, error_code, error_message and error_reason for payment attempt
2. if it's an active attempt, intent status is also updated along with attempt status
3. unified code and message, inferred from GSM for new error_code, error_message and populated in payment attempt


### Additional Changes





## Motivation and Context



## How did you test it?
**Tested Manually**
Payments Manual Update API CURL
```
curl --location --request PUT '{{BASE_URL}}/payments/{{PAYMENT_ID}}/manual-update' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: {{ADMIN_API_KEY}}' \
--data-raw '{
    "attempt_id": {{ATTEMPT_ID}},
    "merchant_id": {{MERCHANT_ID}},
    "attempt_status": {{ATTEMPT_STATUS}}
}'
```

1. Update the Attempt status from failure to charged (active attempt), observe the changes of status in payment_attempt and payment_intent
![image](https://github.com/juspay/hyperswitch/assets/56996463/195fea4f-fa61-4088-846d-f51163be29c4)
![image](https://github.com/juspay/hyperswitch/assets/56996463/63a31f11-4d31-4a17-a558-12cef47c7efd)
![image](https://github.com/juspay/hyperswitch/assets/56996463/2068d0d7-e8c8-467f-b333-aa649d9e8d07)
![image](https://github.com/juspay/hyperswitch/assets/56996463/e073e8ff-5e4e-463e-9f92-37d522f3767b)
![image](https://github.com/juspay/hyperswitch/assets/56996463/c061108b-fac3-4d6e-9054-2486dc7b3bf5)

2. Update the Attempt status from pending to failure (active attempt) along with error code and message, observe the changes of status in payment_attempt and payment_intent
![image](https://github.com/juspay/hyperswitch/assets/56996463/16fd0cd2-07a3-4cfa-aa63-e6a89b98f7bd)
![image](https://github.com/juspay/hyperswitch/assets/56996463/8860780b-c412-4558-82d9-8d1fc6eca8d3)
![image](https://github.com/juspay/hyperswitch/assets/56996463/7b8c934c-95ba-4862-a077-56d769dcc74e)
![image](https://github.com/juspay/hyperswitch/assets/56996463/99e5cc3c-fe79-4581-acaf-ee078a04111f)
![image](https://github.com/juspay/hyperswitch/assets/56996463/77034d5e-0760-4c9b-850b-621ee7ce78fa)

3. Update the Attempt status from pending to failure (inactive attempt), observe the changes of status in payment_attempt and payment_intent (no change in intent status is expected)
![image](https://github.com/juspay/hyperswitch/assets/56996463/28facda3-7628-49d6-8037-7430163bc3ab)
![image](https://github.com/juspay/hyperswitch/assets/56996463/35ed46ad-838d-4615-a678-028535a091a2)
![image](https://github.com/juspay/hyperswitch/assets/56996463/f61b3105-7664-4350-8305-37c330f7c253)
![image](https://github.com/juspay/hyperswitch/assets/56996463/db2360f0-f247-4b55-b6da-9fe56229df15)
![image](https://github.com/juspay/hyperswitch/assets/56996463/35e9b3e3-d480-4e37-a321-c392051f14d1)

4. add unified_code, unified_message in GSM, Update the attempt status from pending to failure and observe the changes of status, unified code and message in payment attempt
![image](https://github.com/juspay/hyperswitch/assets/56996463/5ef2879f-9c43-4174-b863-139e5d3f2af1)
![image](https://github.com/juspay/hyperswitch/assets/56996463/2b87ce3b-8a33-4698-b28d-16a8cb4b1ec4)
![image](https://github.com/juspay/hyperswitch/assets/56996463/1c3b71ad-7a09-400f-88ea-ae9884367391)
![image](https://github.com/juspay/hyperswitch/assets/56996463/25fff31f-220f-4862-8f91-980b7119d8d6)

## Checklist
